File: jiffy4j/core/src/main/java/jiffy/notification/NotificationHandler.java
package jiffy.notification;

import java.nio.ByteBuffer;
import java.util.List;
import java.util.function.Consumer;
import jiffy.notification.event.Control;
import jiffy.notification.event.Error;
import jiffy.notification.event.KVOpKind;
import jiffy.notification.event.Notification;
import jiffy.notification.event.Success;
import jiffy.storage.block_response_service;
import jiffy.storage.sequence_id;
import org.apache.thrift.TException;

public class NotificationHandler implements block_response_service.Iface {

  private Consumer<Notification> notificationConsumer;
  private Consumer<Control> controlConsumer;

  NotificationHandler(Consumer<Notification> notificationConsumer,
      Consumer<Control> controlConsumer) {
    this.notificationConsumer = notificationConsumer;
    this.controlConsumer = controlConsumer;
  }

  @Override
  public void response(sequence_id seq, List<ByteBuffer> result) throws TException {
    throw new TException("NotificationHandler does not support response");
  }

  @Override
  public void chainAck(sequence_id seq) throws TException {
    throw new TException("NotificationHandler does not support chainAck");
  }

  @Override
  public void notification(String op, ByteBuffer data) {
    notificationConsumer.accept(new Notification(new KVOpKind(op), data));
  }

  @Override
  public void control(jiffy.storage.response_type type, List<String> ops, String error) throws TException {
    if (error.isEmpty()) {
      controlConsumer.accept(new Success(ops, type));
    } else {
      controlConsumer.accept(new Error(error, type));
    }
  }
}

File: libjiffy/src/jiffy/storage/notification/notification_handler.cpp
#include <iostream>
#include "notification_handler.h"

namespace jiffy {
namespace storage {

notification_handler::notification_handler(mailbox_t &notifications,
                                           mailbox_t &controls)
    : notifications_(notifications), controls_(controls) {}

void notification_handler::notification(const std::string &op, const std::string &data) {
  notifications_.push(std::make_pair(op, data));
}

void notification_handler::control(response_type type, const std::vector<std::string> &ops, const std::string &error) {
  std::string msg = type == response_type::subscribe ? "Subscribe: " : "Unsubscribe: ";
  if (error.empty()) {
    for (const auto &op: ops)
      msg += op + ", ";
    msg.pop_back();
    msg.pop_back();
    controls_.push(std::make_pair("success", msg));
  } else {
    controls_.push(std::make_pair("error", msg + error));
  }
}

void notification_handler::response(const sequence_id &, const std::vector<std::string> &) {
  throw std::logic_error("Notification handler cannot handle query responses.");
}

void notification_handler::chain_ack(const sequence_id &) {
  throw std::logic_error("Notification handler cannot handle chain acknowledgements.");
}

}
}
